### ISSUE TITLE
Feature: Retry Fetching Images

### ISSUE THREAD

**Issue by dfreilich** [buildpacks/lifecycle#414]
Request from slack [here](https://buildpacks.slack.com/archives/C94UJCNV6/p1600835870001900)

When trying to create 20 paketo images at the same time, 19 succeeded but one failed `to new app image: Get "https://index.docker.io/v2/paketobuildpacks/run/manifests/full-cnb-cf": EOF` . This seems like a temporary network issue, and retry will pass it. Is it possible to add retry step when trying to Get paketobuildpacks from docker.io?

I asked in paketo buildpacks slack, they mentioned maybe you can help me, is there a way to retry getting run image?

My detailed error message is here:
```
[ERROR] Print log for buildrun perf-br-buildpacks-v3-small-15:
2020/09/21 18:10:06 Skipping step because a previous step failed
{"level":"info","ts":1600711726.2768068,"caller":"creds-init/main.go:44","msg":"Credentials initialized."}
{"level":"info","ts":1600711785.0376236,"caller":"git/git.go:139","msg":"Successfully cloned https://github.com/cloudfoundry/cf-acceptance-tests @ b48b42ece754df50d61e745e38ac79c39d5fcee8 (grafted, HEAD, origin/master) in path /workspace/source"}
{"level":"info","ts":1600711787.5540295,"caller":"git/git.go:180","msg":"Successfully initialized and updated submodules in path /workspace/source"}
paketo-buildpacks/node-engine 0.0.262
paketo-buildpacks/npm         0.1.80
Node Engine Buildpack 0.0.262
  Resolving Node Engine version
    Candidate version sources (in priority order):
       -> "*"
    Selected Node Engine version (using ): 10.22.0
  Executing build process
    Installing Node Engine 10.22.0
      Completed in 4.161s
  Configuring environment
    NODE_ENV     -> "production"
    NODE_HOME    -> "/layers/paketo-buildpacks_node-engine/node"
    NODE_VERBOSE -> "false"
    Writing profile.d/0_memory_available.sh
      Calculates available memory based on container limits at launch time.
      Made available in the MEMORY_AVAILABLE environment variable.
NPM Buildpack 0.1.80
  Resolving installation process
    Process inputs:
      node_modules      -> "Not found"
      npm-cache         -> "Not found"
      package-lock.json -> "Not found"
    Selected NPM build process: 'npm install'
  Executing build process
    Running 'npm install --unsafe-perm --cache /layers/paketo-buildpacks_npm/npm-cache'
      Completed in 2.4s
  Configuring environment
    NPM_CONFIG_LOGLEVEL   -> "error"
    NPM_CONFIG_PRODUCTION -> "true"
    PATH                  -> "$PATH:/layers/paketo-buildpacks_npm/modules/node_modules/.bin"
Warning: Warning: analyzed TOML file not found at 'analyzed.toml'
ERROR: failed to new app image: connect to repo store 'index.docker.io/paketobuildpacks/run:full-cnb-cf': Get "https://index.docker.io/v2/paketobuildpacks/run/manifests/full-cnb-cf": EOF
```

**Comment by jromero** [comment]
FWIW, GGCR provides a [retry mechanism](https://github.com/google/go-containerregistry/blob/d448c55e6b46c1d58522a2e4ef0d5a4efda9f28c/pkg/v1/remote/transport/retry.go).

**Comment by ekcasey** [comment]
@jromero I am pretty sure we are already using GGCRs retry mechanism (by default). Here is the discussion from the previous time this issue was raised https://github.com/buildpacks/lifecycle/issues/122

**Comment by jromero** [comment]
I’m curious how reproducible this particular issue is. Ignoring the proposal of a retry mechanism, if the reproduction rate is high, I do wonder if it’s ultimately another issue.

**Comment by EmilyEmily** [comment]
@jromero , from my daily test, it never happened if I only create single paketo image at the time. Occasionally when creating 20 paketo images at the same time, one of the 20 paketo may meet above EOF error when getting index.docker.io. Above are all the error messages I have. 

Could you let me know which lifecycle version includes the #122 retry mechanism? Will there be any output when retry mechanism is triggered?

**Comment by ekcasey** [comment]
Hey @EmilyEmily :)

> Could you let me know which lifecycle version includes the #122 retry mechanism?

The retry mechanism exists in the version you are using (`v0.9.1`). Because we got retries for free with a ggcr update I am not sure in exactly which lifecycle version the retry logic was first included, but suffice it to say its been around for awhile (over a year).

> Could you let me know which lifecycle version includes the #122 retry mechanism? Will there be any output when retry mechanism is triggered?

Unfortunately, no.

>  Is it possible that if for some time network is very bad, and default retry time has been done, but still getting image from docker.io not complete?

I suspect one of two things is happening:
1. The network is very bad and the connection is failing despite the retries (as you suggest).
2. The specific type of error you are seeing is not one that triggers the retries. Only errors that are designated as temporary trigger retries https://github.com/google/go-containerregistry/blob/master/pkg/v1/remote/transport/retry.go#L67

We can attempt to figure out which of the above cases we are in and possibly mitigate in our code or with a PR to ggcr. However this might be difficult depending on how reliably we can reproduce the error. I would like to look into this but, in the interest of full disclosure, it may take some time for this to bubble to the top of the priority queue.

**Comment by EmilyEmily** [comment]
Understood. Thank you @ekcasey for the response, waiting for your investigation.

**Comment by dwillist** [comment]
After poking around this issue for a bit this is what I think is going on

The creation of a remote image in `newV1Image` call [here](https://github.com/buildpacks/imgutil/blob/72f0739080b03c1a09e5bdbec233a8f4c98b9afc/remote/remote.go#L92) uses the `http.DefaultTransport` docs [here](https://golang.org/pkg/net/http/#RoundTripper). When creating a bunch of images at the same time we are attempting to re-use a TCP connection that is closed, which is why we get an EOF error.

### Why

 By design the `http.Transport` tries to re-use existing `TCP` connections, but dockerhub aggressively closes `TCP` connection to `index.docker.io` that have already been used (We have about 4 seconds before a connection is forcefully closed by the server, and `KEEPALIVE` packets are ignored). This is not a problem when we are doing singular builds as most of the traffic to `index.docker.io` is just metadata, but when doing a bunch of builds in parallel it seems like we can hit these timeouts.

### Solutions

The easiest solution is just to simple retry the request on an `io.EOF` error (caused by a socket being closed), but this feels like were solving a networking issue in our application code, which is a bit gross. 

Second option is setting `DisableKeepAlives` on the http.Transport,  which would cause a new connection to be created every time we create an http request. Though this incurs a performance penalty of about 100-300ms every time we need to set up a new TCP connection.

Feels like there might be a better alternative here but I don't have enough context on networking in go to get it.

**Comment by aemengo** [comment]
@dwillist Thanks for the investigation and write-up.

I'll concede that retrying, while normally fine, seems iffy, _in this case_.

I'm leaning towards thinking that this should be handled on the platform side. From what you're saying, DockerHub is presumably doing this to stop you from spamming their APIs. If say, a test suite, is causing this behavior maybe it should do ~10 builds at a time, instead of 20.

Philosophically, I don't know if this issue should drive a change in the lifecycle. Pragmatically? I'm not so sure.

I'm curious what others think.

**Comment by natalieparellano** [comment]
@EmilyEmily lifecycle 0.11.0 will retry image fetches on EOF errors. Would you be able to try it out and let us know if this resolves the issue?

**Comment by natalieparellano** [comment]
Closing this issue as we believe it has been addressed.
